### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by stebogit**
@RaumZeit here we go, I added a few simple tests.
I also took the liberty of adding few input validation for both iso functions (see #9). You might want to add some other validation.

You run the tests with
```
$ npm test
```
and basically all the files in the `test/data/in/` folder are fed to `isoBand` (`isoCountour` still to be set up) and the output compared to the correspondent `test/data/out/` file.

At this point though I believe there is still some issue with `isoBand` output.
For the tests I used a couple of trivial inputs and the array used in the example, which are all OK. I also used few cases I mentioned in #6, and one of them still fails.
Truth is for me is quite difficult to check if the output is actually correct, other than displaying it with the function used in the example.
Would you mind check?
Do you have maybe some datasets you used for your own tests while developing? Would you like to add them as test cases here and see how it goes?

Finally, even though the tests pass, I noticed the `isoBands` output bands/polygons sometimes are not closed. Would you mind check that as well?

### CODE REVIEW

File: test/data/out/issue#6-0_3.json [hunk 1]
```diff
@@ -0,0 +1,9 @@
+[
+  [
+    [0,0],[0,5],[6,5],[6,0]
```
Comment (stebogit) [inline test/data/out/issue#6-0_3.json:3]: this polygon is not closed

File: test/data/out/minimal.json [hunk 2]
```diff
@@ -0,0 +1,9 @@
+[
+  [
+    [0.5,  1],
```
Comment (stebogit) [inline test/data/out/minimal.json:3]: this is not closed either

**Comment by RaumZeit** [comment]
@stebogit Thanks for the PR!
I'll try checking everything this weekend and also find some more complex input data I've been working with...
Just as a side note: I explicitly omit the last coordinate in the `IsoBand` polygons that closes the path by moving to the origin, i.e. the first coordinate in the path. First and last coordinate are identical anyway, and when I use the `IsoBand` polygons, I usually append the `Z` option to the SVG `<path>`. The idea was simply that triangles should be delimited by 3 coordinates, tetragons, by 4, etc...

If first and last coordinate must be identical for any reason, e.g. if this is required in `turfjs`, I can simply add an option to turn-on this behavior.

Note, that only the `IsoContour` (iso line) algorithm may return non-closed paths. The `IsoBand` implementation _always_ returns closed polygons...

**Comment by stebogit** [comment]
@RaumZeit Turf (or better GeoJSON specs) does require a matching first and last coordinates for polygons, while lines _might_ have them matching. If you think about it that makes sense because it is the only way programmatically to be sure (i.e. testable) that what you are dealing with is in fact a closed path. I am used to this convention (and [I'm not alone](#11)), but it looks like SVG instead does not require duplicated coordinates (didn't know about that).
However, in my opinion, _assuming_ the ends of the polygons match is not a reliable practice. If you just need lines, which might or might not be closed, then you already have `IsoContour`(soon to be `IsoLines` 😄) for that. What would be the actual difference in output otherwise?


> Just as a side note: I explicitly omit the last coordinate in the `IsoBand` polygons that closes the path by moving to the origin

Is this a recent change? At the very least this is a breaking change, so since it looks like you are following [semver](https://semver.org/) you should bump the major version.
Alternatively, I'd suggest instead you add something like a `duplicated-off` option and change just the minor version.

In any way though, apparently the output of `IsoBand` is currently not consistent in this regard, returning sometimes duplicated coordinates, sometimes not (see test output).

**Comment by RaumZeit** [comment]
@stebogit 
> Turf (or better GeoJSON specs) does require a matching first and last coordinates for polygons, while lines might have them matching.

I didn't know about the convention for closed paths in GIS. With SVGs and d3.js everything worked fine, and I simply figured since the `IsoBands` implementation should always return closed polygons, I can simply omit the last coordinate and close `<path>` tags using the `Z` option. Anyway, I restored compatibility now for both implementations with v1.2.3! Also, both implementations take care that by default, closed paths with identical start and end coordinates are returned.

> If you just need lines, which might or might not be closed, then you already have IsoContour(soon to be IsoLines ￼) for that. What would be the actual difference in output otherwise?

For me, `IsoContours` always was something that returns iso lines delimiting an area with values lower than `threshold`, while `IsoBands` is an area, i.e. a polygon, that captures values between two thresholds. Again, `IsoBands` never explicitly did, or was intended to return simple iso lines, but always closed paths. With version 1.2.2 I simply removed one of the redundant first coordinates without knowing that this might break compatibility with GIS applications. Still, one could of course produce the iso band polygon(s) from two consecutive iso line computations, but that would take twice the time to compute.


I bumped only the patch version, v1.2.2 clearly broke compatibility, so this version is a bugfix ;)

I'll look into the tests within the next days...

File: marchingsquares-isobands.js [hunk 3]
```diff
@@ -71,6 +71,15 @@
...
     var settings = {};
 
+    /* validation */
+    if (!data) throw new Error('data is required');
+    if (!Array.isArray(data) || !Array.isArray(data[0])) throw new Error('data should be an array of arrays');
+    if (minV === undefined || minV === null) throw new Error('lowerBand is required');
+    if (bandwidth === undefined || bandwidth === null) throw new Error('bandWidth is required');
+    if (isNaN(+minV)) throw new Error('lowerBand must be a number');
+    if (isNaN(+bandwidth)) throw new Error('bandWidth must be a number');
+    if (!!options && options.constructor !== 'object') throw new Error('options must be an object');
```
Comment (RaumZeit) [inline marchingsquares-isobands.js:81]: This test fails, at least in web browser. `options.constructor` is a `function`, so I would suggest using `typeof options !=== 'object'`

File: marchingsquares-isocontours.js [hunk 4]
```diff
@@ -36,6 +36,13 @@
   var settings = {};
 
   function isoContours(data, threshold, options){
+    /* validation */
+    if (!data) throw new Error('data is required');
+    if (!Array.isArray(data) || !Array.isArray(data[0])) throw new Error('data should be an array of arrays');
+    if (threshold === undefined || threshold === null) throw new Error('threshold is required');
+    if (isNaN(+threshold)) throw new Error('threshold must be a number');
+    if (!!options && options.constructor !== 'object') throw new Error('options must be an object');
```
Comment (RaumZeit) [inline marchingsquares-isocontours.js:44]: Again, `options.constructor` is a function, at least in web browser

File: marchingsquares-isobands.js (hunk 3 above)
Comment (stebogit) [reply marchingsquares-isobands.js:81]: Oh, glad you got that, never tested in the browser 👍

**Comment by stebogit** [comment]
> For me, `IsoContours` always was something that returns iso lines delimiting an area with values lower than `threshold`, while `IsoBands` is an area, i.e. a polygon, that captures values between two thresholds. 

I think once `IsoLines` will be in place all will be much clearer to any user. Anyway, specify the behavior of each function in the docs is always best.

> Still, one could of course produce the iso band polygon(s) from two consecutive iso line computations, but that would take twice the time to compute.

I did try that, but the issue there is that this way you programmatically can not (or at least not in an easy way) reliably figure out which line contains which, so you can identify the area inside them, or which is the external line for a particular area, especially when you have multiple lines defining holes inside a band. In GeoJSON at least, that is necessary again to [define polygons](https://tools.ietf.org/html/rfc7946#section-3.1.6), where the external linearRing must be the first in the array.

**Comment by RaumZeit** [comment]
> I think once `IsoLines` will be in place all will be much clearer to any user ...

That's right! As soon as the tests are complete and merged, I'll add the `IsoLines` interface and bump the minor version. But I think it would be best to keep the `IsoContour` function for backward compatiblity  reasons, until more features have accumulated that justify an increase of the major version.

**Comment by stebogit** [comment]
> I think it would be best to keep the `IsoContour` function for backward compatibility reasons

definitely! 👍

**Comment by RaumZeit** [comment]
@stebogit Would you please have a look at the [dev_modules](https://github.com/RaumZeit/MarchingSquares.js/tree/dev_modules) branch? I've changed the source tree and formatted the library in ES2015, using `rollup.js` to bundle the final UMD distribution files. It's not finished yet, since now that I use `rollup.js` the minified scripts can be create automatically as well (until now, I've been uploading the original sources to some closure compiler web interface).
Also, I've already squash-merged your `test` branch and changed the expected output for all but one test case. 

Btw.: How did you create the expected output json files? Did you manually computed the expected coordinates?

**Comment by stebogit** [comment]
Sure @RaumZeit, I'll be happy to.
Would you mind though creating a PR for that branch (and maybe connecting it #8 for better traceability)? That way is gonna be easier for me to add comments.

BTW funny I had already started working on the same issue [here](https://github.com/RaumZeit/MarchingSquares.js/compare/master...stebogit:esm), but yours seems better! 😄

**Comment by stebogit** [comment]
>Btw.: How did you create the expected output json files? Did you manually computed the expected coordinates?

I just modified the `example.html` a bit to `log()` the calculated band I needed when feeding the input data I wanted to test. A bit clunky but that way I could also see the rendered output.

**Comment by RaumZeit** [comment]
> BTW funny I had already started working on the same issue here, but yours seems better!

That's funny, I didn't know that! I was looking for how to provide ES2015 modules, and then came across `rollup.js`. It seemed to make things pretty easy, especially still providing UMD wrapped modules while developing with the ES2015 modules standard...
I've create a PR as you requested... if you have anything that doesn't look like javascript standard, feel free to add review comments! I'm more a `C` guy, so I don't know how to best format things for `js`.

**Comment by RaumZeit** [comment]
I close this PR in favor of #14

**Pull request closed without merging**